### PR TITLE
Do not append parent more than once when calling serialize.

### PR DESCRIPTION
**Pull request by Yannig**
Can result in a 10 time speed improvement when using role with redundant dependencies (check https://github.com/ansible/ansible/issues/14112 for more details).

### CODE REVIEW

File: lib/ansible/playbook/role/__init__.py [hunk 1]
```diff
@@ -376,7 +376,10 @@ def serialize(self, include_deps=True):
 
         parents = []
         for parent in self._parents:
-            parents.append(parent.serialize(include_deps=False))
+            # include parent only once
+            # Fix https://github.com/ansible/ansible/issues/14112
+            if parent not in self._parents:
```
Comment (jimi-c) [inline lib/ansible/playbook/role/__init__.py]: Should this be `if parent not in parents:` instead? Otherwise, I do not see how this helps anything.

Comment (Yannig) [inline lib/ansible/playbook/role/__init__.py]: @jimi-c Well, you're right ... This condition is always false so it's quite normal to get a huge speed improvement.

Comment (Yannig) [inline lib/ansible/playbook/role/__init__.py]: But it's strange to see that there's no regression in test suite.

**Comment by Yannig** [comment]
@jimi-c this time the commit is a little bit smarter (which is not difficult indeed ...). I'm using an array of string to keep track of already included role.

**Comment by Yannig** [comment]
The improvement is still important (1s vs 14s without this patch).

**Comment by thiagocotting** [comment]
Do we have an estimative when this will be merged? It's really painful to run roles with dependencies without this change

**Comment by jimi-c** [comment]
@Yannig in my performance branch, we actually don't even serialize anymore: https://github.com/ansible/ansible/compare/performance_improvements

**Comment by Yannig** [comment]
No more relevant in regard of Ansible performance in 2.2.

**Pull request closed without merging**